base/special/cbrt.jl

```diff
@@ -0,0 +1,121 @@
+# ====================================================
+# Copyright (C) 1993 by Sun Microsystems, Inc. All rights reserved.
+#
+# Developed at SunPro, a Sun Microsystems, Inc. business.
+# Permission to use, copy, modify, and distribute this
+# software is freely granted, provided that this notice
+# is preserved.
+# ====================================================
+#
+# Optimized by Bruce D. Evans.
+
+# s_cbrtf.c -- float version of s_cbrt.c.
+# Conversion to float by Ian Lance Taylor, Cygnus Support, ian@cygnus.com.
+# Debugged and optimized by Bruce D. Evans.
+
+cbrt(x::Real) = cbrt(float(x))
+
+issubnormal(::Float64, hw) = hw < 0x00100000
+issubnormal(::Float32, hw) = hw < 0x00800000
+
+const cbrt_B1_32 = 0x2a5119f2 # UInt32(709958130) # B1 = (127-127.0/3-0.03306235651)*2**23
+const cbrt_B1_64 = 0x2a9f7893 # UInt32(715094163) # B1 = (1023-1023/3-0.03306235651)*2**20
+
+const cbrt_B2_32 = 0x265119f2 # UInt32(642849266) # B2 = (127-127.0/3-24/3-0.03306235651)*2**23
+const cbrt_B2_64 = 0x297f7893 # UInt32(696219795) # B2 = (1023-1023/3-54/3-0.03306235651)*2**20
+
+cbrt_t(::Type{Float64}) = 1.8014398509481984e16# 2.0^54
+cbrt_t(::Type{Float32}) = 1.6777216f7 ## 2f0^24
+
+cbrt_t_from_words(::Type{Float32}, s, high)   = reinterpret(Float32, s | (div(high & 0x7fffffff, 0x00000003))+cbrt_B2_32)
+cbrt_t_from_words_alt(::Type{Float32}, s, hx) = reinterpret(Float32, s | (div(hx, 0x00000003) + cbrt_B1_32))
+
+cbrt_t_from_words(::Type{Float64}, s, high)   = reinterpret(Float64, UInt64(s | (div(high & 0x7fffffff, 0x00000003)+cbrt_B2_64))<<32)
+cbrt_t_from_words_alt(::Type{Float64}, s, hx) = reinterpret(Float64, UInt64(s | (div(hx, 0x00000003) + cbrt_B1_64))<<32)
+
+function cbrt(x::Tf) where Tf <: Union{Float32, Float64}
+    # mathematically cbrt(x) is defined as the real number y such that y^2 == x
+
+    hw = highword(x) % Int32
+    s = hw & 0x80000000
+    hw = xor(hw, s)
+
+    if isnan(x) || isinf(x)
+        return x
+    end
+
+    # rough cbrt to 5 bits
+
+    #    cbrt(2**e*(1+m) ~= 2**(e/3)*(1+(e%3+m)/3)
+    # where e is integral and >= 0, m is real and in [0, 1), and "/" and
+    # "%" are integer division and modulus with rounding towards minus
+    # infinity.  The RHS is always >= the LHS and has a maximum relative
+    # error of about 1 in 16.  Adding a bias of -0.03306235651 to the
+    # (e%3+m)/3 term reduces the error to about 1 in 32. With the IEEE
+    # floating point representation, for finite positive normal values,
+    # ordinary integer divison of the value in bits magically gives
+    # almost exactly the RHS of the above provided we first subtract the
+    # exponent bias and later add it back.  We do the
+    # subtraction virtually to keep e >= 0 so that ordinary integer
+    # division rounds towards minus infinity; this is also efficient.
+
+    if issubnormal(x, hw) # zero or subnormal?
+        if x == zero(Tf)
+            return x # cbrt(+-0) is itself
+        end
+        t = x*cbrt_t(Tf)
```

Comment:
Couldn't this overflow if `x` is close to `realmax(Tf)`?

Reply:
sure, but `x` is subnormal here